spec/index.bs

```diff
@@ -9342,7 +9397,61 @@ The determination is based on |descriptor|.{{GPURenderPipelineDescriptor/multisa
 
-Issue: fill out this section
+Fragment processing stage is a programmable stage of the render [=pipeline=] that
+evaluates the fragment data (generally, a color) to be written into render targets.
+
+This stage produce a <dfn dfn>Fragment</dfn> for each [=RasterizationPoint=]:
+<div algorithm="Fragment accessors" dfn-for=Fragment>
+    - <dfn dfn>destination</dfn> refers to [=FragmentDestination=].
+    - <dfn dfn>coverageMask</dfn> refers to multisample coverage mask (see [[#sample-masking]]).
+    - <dfn dfn>depth</dfn> refers to the depth in [=NDC=] coordinates.
+    - <dfn dfn>colors</dfn> refers to the list of color values,
+        one for each target in {{GPURenderPassDescriptor/colorAttachments}}.
+</div>
+
+<div algorithm>
+    <dfn abstract-op>process fragment</dfn>(rp, desc, state)
+
+    **Arguments:**
+    - |rp|: The [=RasterizationPoint=], produced by [[#rasterization]].
+    - |desc|: The descriptor of type {{GPUFragmentState}}.
+    - |state|: The active [=RenderState=].
+
+    **Returns:** [=Fragment=] or `null`.
+
+    1. Let |fragment| be a new [=Fragment=] object.
+    1. Set |fragment|.[=Fragment/destination=] to |rp|.[=RasterizationPoint/destination=].
+    1. Set |fragment|.[=Fragment/coverageMask=] to |rp|.[=RasterizationPoint/coverageMask=].
+    1. Set |fragment|.[=Fragment/depth=] to |rp|.[=RasterizationPoint/depth=].
+    1. If |desc| is not `null`:
+      1. Set the shader input [=builtins=]:
+          - Set the `position` builtin, if any, to
+              `vec4<f32>`(|rp|.[=RasterizationPoint/destination=].[=FragmentDestination/position=], |rp|.[=RasterizationPoint/depth=], |rp|.[=RasterizationPoint/perspectiveDivisor=]).
+          - Set the `front_facing` builtin, if any, to |rp|.[=RasterizationPoint/frontFacing=].
+          - Set the `sample_index` builtin, if any, to |rp|.[=RasterizationPoint/destination=].[=FragmentDestination/sampleIndex=].
+          - Set the `sample_mask` builtin, if any, to |rp|.[=RasterizationPoint/coverageMask=].
+      1. For each user-specified [=pipeline input=] of the fragment stage:
+        1. Let |value| be the interpolated fragment input,
+            based on |rp|.[=RasterizationPoint/barycentricCoordinates=], |rp|.[=RasterizationPoint/primitiveVertices=],
+            and the [=interpolation=] qualifier on the input.
+            Issue: describe the exact equations.
+        1. Set the corresponding fragment shader [=location=] input to |value|.
+      1. Invoke the fragment shader entry point described by |desc|.
+      1. If the fragment issued `discard`, return `null`.
+      1. Set |fragment|.[=Fragment/colors=] to the user-specified [=pipeline output=] values from the shader.
+      1. Take the shader output [=builtins=]:
+        1. If `frag_depth` [=builtin=] is produced by the shader as |value|:
+          1. Let |vp| be |state|.[=RenderState/viewport=].
+          1. Set |fragment|.[=Fragment/depth=] to clamp(|value|, |vp|.`minDepth`, |vp|.`maxDepth`).
```

Comment:
There will be a merge conflict with #2139 because I landed that today. This line takes care of clamping frag_depth, but we also need to clamp if unclippedDepth=true and frag_depth is not used by the shader.